### ISSUE TITLE
CoroutineExceptionHandler doesn't receive exception thrown after cancellation

### ISSUE THREAD

**Issue by kptlronyttcna** [Kotlin/kotlinx.coroutines#356]
`exceptionHandler` below only receives one `JobCancellationException` and doesn't receive `RuntimeException`. Is this the expected behavior? If so, what happens to `RuntimeException`?

```
val exceptionHandler = CoroutineExceptionHandler { _, t ->
  println("got $t")
}

val job = launch(CommonPool) {
  launch(coroutineContext + exceptionHandler) {
    Thread.sleep(1000)
    throw RuntimeException("OH OH")
  }
  delay(100000000)
}

delay(100)
job.cancelAndJoin()
```

**Comment by kptlronyttcna** [comment]
> because the RuntimeException hasn't be thrown yet, you get the JobCancellationException

But isn't coroutine cancellation cooperative? I mean there isn't anything cancellable (like yield) between `Thread.sleep` and `throw RuntimeException`, so even if the job is cancelled, the `RuntimeException` will be thrown. Wouldn't that (if not catched) make the thread fail?

**Comment by jcornaz** [comment]
Yes the cancellation is cooperative and `RuntimeException` will be always thrown. It will just be ignored.

As soon as the coroutine is cancelled it calls the exception handler passing the Cancellation exception, even if the coroutine is actually still running. And because an exception has already been handled for this coroutine, the next one (`RuntimeException`) is ignored. It will not make the thread fail, the exception is just lost.

If it is a good design choice is arguable. I agree that we should always try to not lose exception.

However if we consider that:
1. Good use of coroutines generally use cancellable suspend functions. And usually long running coroutines are cooperatively cancellable.
2. From the exception handler view, it doesn't really matter. The important think to know is that the coroutine failed. The sooner we know it, the better. And as we know that the coroutine has been cancelled, there is no point to wait, we can directly considered the coroutines as failed because of cancellation, even if it is in practice still running (maybe releasing resources, or whatsoever). But it wouldn't make sense to handle two exception for the same coroutine, would it?

Actually, may I ask if you have an issue due to this behavior?

**Comment by qwwdfsad** [comment]
>If it is a good design choice is arguable.

Note that this choice is being re-worked right now: #333
Behaviour reported by @kptlronyttcna should be at least well documented and **maybe** reworked to avoid exception loss. Target `RuntimeException` should either be passed to `ExceptionHandler` or added to JCE cause/suppressed exceptions

**Comment by kptlronyttcna** [comment]
> Actually, may I ask if you have an issue due to this behavior?
I am mostly confused. For example the second example I gave (the one with no child coroutine) is not consistent with the first one.

**Comment by jcornaz** [comment]
>For example the second example I gave (the one with no child coroutine) is not consistent with the first one.

Good point, it should be consistent. Let's follow #333

**Comment by qwwdfsad** [comment]
Will be fixed under #333
